Feature/fcm implementation

RamonKea-inh opened the pull request.

RamonKea-inh committed:
Added FCM client, but might use the SDK instead

RamonKea-inh committed:
Added models

RamonKea-inh committed:
added Postgres db schema

RamonKea-inh committed:
added config for the fcm client

RamonKea-inh committed:
added postgres db script

RamonKea-inh committed:
added message endpoints

RamonKea-inh committed:
Added middleware

RamonKea-inh committed:
Added token CRUD

RamonKea-inh committed:
Added notification logic

RamonKea-inh committed:
Update main.go

RamonKea-inh committed:
Revert "added postgres db script"

This reverts commit 59dbd3ca08693c1bd4245f55389ef16881772458.

RamonKea-inh committed:
Revert "Added middleware"

This reverts commit a853418f38fc534eb402d7b7687271711b7c31ff.

RamonKea-inh committed:
Revert "added Postgres db schema"

This reverts commit 2a0d7e3c756faa73e9d5228955bf8c7856e0ce93.

RamonKea-inh committed:
changed the fcm client to use the sdk and repositories to use azure table storage

RamonKea-inh committed:
Adjusted the config and README file to set the right needs

RamonKea-inh committed:
Update config.go

RamonKea-inh committed:
adapted the notification service to not use the fcm client and be less dependent

RamonKea-inh committed:
Update .gitignore

RamonKea-inh committed:
Update message_repository.go

RamonKea-inh committed:
Update group_messages_controller.go

RamonKea-inh committed:
Update message_service.go

RamonKea-inh committed:
Made repository for saving FCMTokens

RamonKea-inh committed:
Refactoring and interfaces

- Changed models
- Changed directory names
- Added interfaces
- Refactored a bunch off methods to better handle the pagination

RamonKea-inh committed:
Delete types.go

made a model for this in models.go

RamonKea-inh committed:
Removed debug code

RamonKea-inh committed:
removed unnecessary azure variables

RamonKea-inh committed:
replaced mux with gin

RamonKea-inh committed:
added Prometheus

RamonKea-inh committed:
Added CORS

RamonKea-inh committed:
Merge branch 'development' into feature/fcm-implementation

RamonKea-inh committed:
Update main.go

added error messages

Tayam-El-Jebari commented on configs/.env.example:
why is this in configs? Root folder is default right?

Tayam-El-Jebari commented on docs/README.md:
```diff
@@ -16,16 +17,18 @@ Groupchat service for PTSS Support, built with Go and FCM
-   jwt_secret: "my-secret-key"
-   fcm_server_key: "your-fcm-server-key"
+3. Configure environment variables:
+   Create a `.env` file in the root directory of the project using the provided `.env.example` file as a reference. Example:
```
.env.example is in configs which also implies it should be in root, please put .env.example then in the same folder as you expect .env to be in.

Tayam-El-Jebari commented on internal/controllers/group_messages_controller.go:
```diff
@@ -0,0 +1,135 @@
+	}
+}
+
+// GetMessages handles GET /groups/messages
+// GetMessages handles GET /groups/messages
```
double comment, also quite unnecessary, would just remove them both
```suggestion
```

Tayam-El-Jebari commented on internal/controllers/group_messages_controller.go:
```diff
@@ -0,0 +1,135 @@
+	"Groupchat-Service/internal/services"
+)
+
+// FCMMessageController handles HTTP requests for messages
```
```suggestion
```

Tayam-El-Jebari commented on internal/controllers/group_messages_controller.go:
```diff
@@ -0,0 +1,135 @@
+	// Parse pagination query
+	query := models.PaginationQuery{
+		PageSize:  10, // Default page size
+		Direction: "next",
```
Direction is a string but can only be next or previous. Change this to an enum.

Tayam-El-Jebari commented on internal/controllers/group_messages_controller.go:
```diff
@@ -0,0 +1,135 @@
+		Direction: "next",
+	}
+
+	if size := ctx.Query("pageSize"); size != "" {
+		pageSize, err := strconv.Atoi(size)
+		if err != nil || pageSize < 1 || pageSize > 50 {
+			ctx.JSON(http.StatusBadRequest, gin.H{"error": "page size must be between 1 and 50"})
+			return
+		}
+		query.PageSize = pageSize
+	}
+
+	if cursor := ctx.Query("cursor"); cursor != "" {
+		query.Cursor = &cursor
+	}
+
+	if direction := ctx.Query("direction"); direction != "" {
+		if direction != "next" && direction != "previous" {
+			ctx.JSON(http.StatusBadRequest, gin.H{"error": "direction must be 'next' or 'previous'"})
+			return
+		}
+		query.Direction = direction
+	}
+
+	if search := ctx.Query("search"); search != "" {
+		if len(search) > 100 {
+			ctx.JSON(http.StatusBadRequest, gin.H{"error": "search term too long: maximum 100 characters"})
+			return
+		}
+		query.Search = &search
+	}
+
```
all these checks make your controller bloated, and seeing as how it's business logic, i think its best to move this to either a service layer, or a util method that would check the entire response.
avoid magic numbers. Put 100 and 50 in constants. They're fine for things like this as the context is explained just before:
``` PageSize: 10 ```

RamonKea-inh replied:
From what I know the checks/validation isn't seen as buisiness logic, but I'll seperate it for you ;)

Tayam-El-Jebari commented on internal/controllers/group_messages_controller.go:
```diff
@@ -0,0 +1,135 @@
+		query.Search = &search
+	}
+
+	// Get messages from service
```
don't need this comment, please go through your code and remove them as it makes the code more cluttered. For everything that isn't explained by the method name (e.g. GetMessages) move it to another method that explains what it does. If you can't do that because a. it's specific logic/ math/ a workaround for a bug/ anything of the sort, then add comments.

Tayam-El-Jebari commented on internal/database/repositories/fcm_token_repository.go:
```diff
@@ -0,0 +1,77 @@
+	_, err = r.table.AddEntity(ctx, marshaled, nil)
+	if err != nil {
+		return fmt.Errorf("failed to save token: %w", err)
+	}
+
+	return nil
+}
+
+func (r *FcmTokenRepository) DeleteToken(ctx context.Context, groupID uuid.UUID, userID uuid.UUID) error {
+	// TODO implement
+
+	return nil
```
we don't need a delete, no?

RamonKea-inh replied:
I still need to create endpoints for the tokens as I later discovered this is also necessarry for FCM tokens. 
Delete token is necessary for the following reasons:  

1. Token Expiry: FCM tokens can expire or become invalid. When this happens, the client should inform the server to remove the old token.
2. User Logout: When a user logs out, their FCM token should be removed from the server to prevent sending notifications to a user who is no longer logged in.
3. Device Change: If a user changes devices, the old device's token should be deleted to avoid sending notifications to a device that the user no longer uses.
4. Token Management: Proper token management helps in maintaining an accurate list of active tokens, ensuring that notifications are sent only to valid and active devices.

Tayam-El-Jebari commented on internal/services/notification_service.go:
```diff
@@ -0,0 +1,143 @@
+		"type":       "group_message",
+	}
+
+	badgeNumber := 1
```
why is badgenumber 1 here? don't you want it to represent the amount of unread messages?

RamonKea-inh replied:
I haven't written the logic for this yet. Will add it.

RamonKea-inh committed:
Update .env.example

RamonKea-inh committed:
Applied some suggested changes

RamonKea-inh committed:
Added constants and removed some comments

RamonKea-inh committed:
Seperated validatoin logic from controller

RamonKea-inh committed:
Added code so the existing storage table will be used if it was already created

SemPlaatsman commented on .env.example:
```diff
@@ -0,0 +1,9 @@
+FIREBASE_CREDENTIAL_FILE=./config/firebase-credentials-dev.json
+
+# Azure Storage Configuration
+AZURE_CONNECTION_STRING=devconnectionstring
```
Is dit niet `UseDevelopmentStorage=true` ofzo?

RamonKea-inh replied:
In principe zou dat ook moeten werken, maar bij mij niet haha.
Ik moet de default dev connection string gebruiken die exact hetzelfde doet en geen speciale"/secret waardes heeft (want het is voor iedereen dezelfde account key).
Dus ik zal het naar dit veranderen: "DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;BlobEndpoint=http://127.0.0.1:10000/devstoreaccount1;QueueEndpoint=http://127.0.0.1:10001/devstoreaccount1;TableEndpoint=http://127.0.0.1:10002/devstoreaccount1;"

SemPlaatsman commented on cmd/main.go:
```diff
@@ -1,5 +1,104 @@
+	router.Use(middleware.PrometheusMiddleware())
+
+	// Define routes
+	router.GET("/groups/:groupId/messages", messageController.GetMessages)
+	router.POST("/groups/:groupId/messages", messageController.CreateMessage)
+	router.PUT("/groups/:groupId/messages/:messageId/pin", messageController.ToggleMessagePin)
```
Misschien beter om de routes in de controller zelf te configuren met een routes method (zie de authentication service)

SemPlaatsman commented on internal/services/notification_service.go:
Als ik dit nu goed begrijp zit de notification service (of tenminste een deel daarvan), nu in de groupchat service gebakken. Waarom is deze keuze gemaakt?

RamonKea-inh replied:
Omdat FCM er ook juist voor zorgt dat je die notificaties kan sturen. Ik dacht eerst ook dat je het nodig had om gewoon die chat te laten werken, maar zo als ik het later pas begreep is het om push notificaties the sturen wanneer apparaten niet meteen bereikbaar zijn en wat andere use cases. Wat ook wel handig is voor de groupchat, omdat dit ook om belangrijke berichten kan gaan.

SemPlaatsman replied:
Yeah i know, maar dit deel wat je nu beschrijft was eerst de rol van de notification service. Als de groupchat service enige is die notifications stuurt is dit geen probleem, maar als meerdere services notifications sturen kan dit beter in een notification service (of ook gwn niet want tijd enzo XD). Eigenlijk hoorde het volledige FCM gedeelte in de notification service, en dan roepte je die gwn aan.

RamonKea-inh replied:
Ja klopt

SemPlaatsman reviewed (changes requested):
Check ook nog even de CodeQL concerns

RamonKea-inh committed:
added logic for getting the badge number

RamonKea-inh committed:
Added interface for the validation service

RamonKea-inh committed:
Added the default connection string for emulating using azurite as example

RamonKea-inh committed:
Update validation_service.go

RamonKea-inh committed:
moved the router config to the controller

RamonKea-inh committed:
Added a validation service and FCM token controller

RamonKea-inh committed:
Fixed a couple bugs that happened when posting and getting messages

SemPlaatsman commented on .env.example:
```diff
@@ -0,0 +1,9 @@
+# Azure Storage Configuration
+AZURE_CONNECTION_STRING=DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;BlobEndpoint=http://127.0.0.1:10000/devstoreaccount1;QueueEndpoint=http://127.0.0.1:10001/devstoreaccount1;TableEndpoint=http://127.0.0.1:10002/devstoreaccount1;
+
+# Application Configuration
+APP_ENV=development
+APP_PORT=8080
```
Wrm is dit nodig? Quarkus regelt dat toch al built-in?

RamonKea-inh replied:
Ik gebruik toch geen Quarkus voor Go?

SemPlaatsman replied:
GVD dit is nou als de tweede keer dat ik dit vergeet :rage3:

Tayam-El-Jebari replied:
🤡

SemPlaatsman commented on internal/models/models.go:
```diff
@@ -0,0 +1,55 @@
+	ID         uuid.UUID `json:"id"`
+	GroupID    uuid.UUID `json:"groupId"`
+	SenderID   uuid.UUID `json:"senderId"`
+	SenderName string    `json:"senderName"`
```
SenderName wordt uit de user service gehaald toch? Anders moet je de naam op twee plaatsen veranderen en dat is zeer bad practice

RamonKea-inh replied:
Ja het wordt nu uit de user service gehaald en daarna opgeslagen bij elke message in de table storage. Dus je bedoelt dat ik de SenderName niet moet opslaan in de Table Storage en de user service voor elke message moet aanroepen om de senderName te krijgen wanneer er een GET voor de messages wordt gedaan? Of gewoon helemaal geen senderName terug sturen en de frontend de senderName laten requesten voor elke message? Lijkt me allebei ook niet helemaal juist.

SemPlaatsman replied:
Is nu een consensus voor. Sla de name en de id van de sender op die je dan uit de access token zou moeten kunnen halen. En dan stuur je terug wat je wilt, want je hoeft niet de sender id terug te sturen als dat niet hoeft, is meer voor auditing en future extension. Daan gaat hetzelfde doen.

RamonKea-inh replied:
welke access token?

SemPlaatsman replied:
Die in de cookies van een ingelogde gebruiker staat, vraag anders maar in de Discord of lees de berichten

SemPlaatsman reviewed (changes requested):
Heb wat clarifications nodig, verder goed

Tayam-El-Jebari commented on internal/controllers/fcm_token_controller.go:
would've been a lot cleaner to use facade pattern here 👀 , but this is fine as well. Just a chunky controller

Tayam-El-Jebari commented on internal/controllers/fcm_token_controller.go:
```diff
@@ -0,0 +1,72 @@
+)
+
+type fcmTokenController struct {
+	service           services.FCMTokenService
```
pls rename this to fcmTokenService for clarity

RamonKea-inh replied:
Rename what to fcmTokenService? FCMTokenService?

SemPlaatsman replied:
```go
type fcmTokenController struct {
	fcmTokenService       services.FCMTokenService
```

vlgns mij

Tayam-El-Jebari commented on internal/database/repositories/message_repository.go:
```diff
@@ -0,0 +1,296 @@
+	pager := r.table.NewListEntitiesPager(&aztables.ListEntitiesOptions{
+		Filter: &filter,
+	})
+
```
why use newlistentitespager? change this to  r.table.GetEntity as it should just get one message back, not a list

Tayam-El-Jebari commented on internal/database/repositories/message_repository.go:
```diff
@@ -0,0 +1,296 @@
+	return nil, fmt.Errorf("message not found")
+}
+
+func (r *messageRepository) GetMessages(ctx context.Context, groupID uuid.UUID, query models.PaginationQuery) ([]models.Message, *models.PaginationResponse, error) {
```
this method is waaaaay too long, please split this up into more digestable methods. Easier testing & more readable code for not only yourself, but also us 🥺

Tayam-El-Jebari commented on internal/database/repositories/message_repository.go:
```diff
@@ -0,0 +1,296 @@
+		Top:    &pageSize,
+	}
+
+	if query.Cursor != nil && *query.Cursor != "" {
```
why
```
	 && *query.Cursor != ""
```
? any specific reason in to treat an empty string cursor differently from a nil cursor?

RamonKea-inh replied:
From what I have gathered:
A nil cursor indicates that no cursor was provided, meaning pagination should start from the beginning.
An empty string cursor ("") might be treated as an invalid or special case, and it should not be used for pagination.

Tayam-El-Jebari commented on internal/models/models.go:
```diff
@@ -0,0 +1,55 @@
+	IsPinned   bool      `json:"isPinned"`
+}
+
+type MessageCreate struct {
+	Content string `json:"content" validate:"required,min=1,max=1000"`
+}
+
+type PaginationQuery struct {
+	Cursor    *string   `json:"cursor,omitempty"`
+	PageSize  int       `json:"pageSize"`
+	Direction Direction `json:"direction"`
+	Search    *string   `json:"search,omitempty"`
+}
+
+type PaginatedResponse struct {
+	Data       interface{}        `json:"data"`
+	Pagination PaginationResponse `json:"pagination"`
+}
+
+type PaginationResponse struct {
+	NextCursor     *string `json:"nextCursor,omitempty"`
+	PreviousCursor *string `json:"previousCursor,omitempty"`
+	HasNext        bool    `json:"hasNext"`
+	HasPrevious    bool    `json:"hasPrevious"`
+}
```
all these need to be moved to a new file, maybe you can put paginationquery and direction in the same file as they are tightly coupled, but rest i would recommend just doing in a different file

RamonKea-inh replied:
For a project this size, this seemed unnecessary to me. But I'll do it if you insist on it...

Tayam-El-Jebari commented on internal/database/repositories/message_repository.go:
```diff
@@ -0,0 +1,296 @@
+	filter := fmt.Sprintf("PartitionKey eq '%s'", groupID.String())
+
+	// Add search filter if provided
+	if query.Search != nil && *query.Search != "" {
+		filter += fmt.Sprintf(" and Content ne '' and Content containing '%s'", *query.Search)
```
maybe add a max size to search in your service?

RamonKea-inh replied:
I have already done this in the validation service, current limit is 100 characters. Do you also reckon thats enough?: 
`if search, ok := queryParams["search"]; ok {
		if len(search) > MaxSearchLength {
			return query, errors.New("search term too long: maximum 100 characters")
		}
		query.Search = &search
	}`

Tayam-El-Jebari commented on internal/database/repositories/message_repository.go:
```diff
@@ -0,0 +1,296 @@
+
+func (r *messageRepository) GetLastReadTime(ctx context.Context, groupID uuid.UUID, userID uuid.UUID) (time.Time, error) {
+	filter := fmt.Sprintf("PartitionKey eq '%s' and RowKey eq '%s'", groupID.String(), userID.String())
+	pager := r.table.NewListEntitiesPager(&aztables.ListEntitiesOptions{
```
use GetEntity here as well pls

Tayam-El-Jebari commented on internal/database/repositories/message_repository.go:
```diff
@@ -0,0 +1,296 @@
+	return &s
+}
+
+func (r *messageRepository) CountUnreadMessages(ctx context.Context, groupID uuid.UUID, userID uuid.UUID, lastReadTime time.Time) (int, error) {
+	filter := fmt.Sprintf("PartitionKey eq '%s' and SentAt gt '%s'", groupID.String(), lastReadTime.UTC().Format(time.RFC3339))
+	pager := r.table.NewListEntitiesPager(&aztables.ListEntitiesOptions{
+		Filter: &filter,
+	})
+
+	unreadCount := 0
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return 0, fmt.Errorf("failed to list unread messages: %w", err)
+		}
+		unreadCount += len(page.Entities)
+	}
+
+	return unreadCount, nil
+}
```
same as previous, you can just change the filter instead of going through all the messages in total

RamonKea-inh replied:
I don't understand your comment, I am currently filtering by the group ID. I still need to count multiple messages to know how many messages aren't read. I can optimize it by only selecting the group ID field in the query, so will add that...

Tayam-El-Jebari commented on internal/services/message_service.go:
```diff
@@ -1,9 +1,94 @@
+		ID:         uuid.New(),
+		GroupID:    groupID,
+		SenderID:   userID,
+		SenderName: userName,
```
shouldn't username be retrieved by this service calling the user service sem made?

RamonKea-inh replied:
Read the earlier conversation with Sem about this please. I am gonna retrieve it through his user service, but that implementation has not been made yet.

SemPlaatsman replied:
The consensus me and Daan came up with was that the SenderName and SenderId can be gathered from the access token, since these cannot be changed, and if further extension of the application means the SenderName could be changed, the transition would be relatively easy to implement due to the SenderId also being present.

RamonKea-inh committed:
refactored get message by id and toggle message pin methods

RamonKea-inh committed:
Refactored/optimized two methods for the notification badge number

RamonKea-inh committed:
Fixed searching in messages

Discovered that Azure Table Storage doesn't have the function to partially search in fields that are not the partition key or row key. So from what I found, the best way to go around this is retrieving all the messages into the memory and searching in the application. Or we need to choose another database to store messages.

RamonKea-inh committed:
Made the methods in the message repository more maintainable

RamonKea-inh committed:
Refactored some code that was redundant into a function

RamonKea-inh committed:
Divided the models into several files

RamonKea-inh committed:
Update fcm_token_controller.go

RamonKea-inh committed:
Added validation for tokens and refactored some other validation logic

RamonKea-inh committed:
Added JWT middleware and refactored the functions to retrieve group ID and user ID through the JWT

RamonKea-inh committed:
Added the retrieval off group member names from the user service for getting group messages

RamonKea-inh committed:
removed sender name from the database

RamonKea-inh committed:
updated env example and fixed a warning

RamonKea-inh committed:
Refactored notification service to be more readable and maintainable

SemPlaatsman commented on internal/middleware/jwt_middleware.go:
Token komt van cookies, niet Bearer. Gebruik `ACCESS_TOKEN_COOKIE_NAME` uit de env (wordt automatisch geinject in openshift) voor de naam van de cookie. Gooi anders gwn de bestaande code uit de base repo in Claude en vraag of je middleware in Go kan maken.

SemPlaatsman reviewed (changes requested):
Nog een jwt middleware change. En ik ga er ook vanuit dat je niet langer de username opslaat maar de contents van de token gebruikt voor een user service call. Dit is alles wat je uit de token mag halen:
```kotlin
data class UserContext(
    val userId: UUID,
    val groupId: UUID?, // Could be empty for Role.ADMIN and Role.HCP
    val roles: Set<Role>,
    val hasPin: Boolean
)
```

RamonKea-inh committed:
Changed JWT middleware to retrieve JWT from cookie instead off bearer token

RamonKea-inh committed:
Update .env.example

RamonKea-inh committed:
Updated the env variables in config
